vault/raft.go

```diff
@@ -573,23 +602,73 @@ func (c *Core) InitiateRetryJoin(ctx context.Context) error {
 }
 
 func (c *Core) JoinRaftCluster(ctx context.Context, leaderInfos []*raft.LeaderJoinInfo, nonVoter bool) (bool, error) {
-	raftStorage, ok := c.underlyingPhysical.(*raft.RaftBackend)
-	if !ok {
-		return false, errors.New("raft storage not configured")
+	raftBackend := c.getRaftBackend()
+	if raftBackend == nil {
+		return false, errors.New("raft backend not in use")
 	}
 
-	if raftStorage.Initialized() {
-		return false, errors.New("raft storage is already initialized")
-	}
+	// NOTE: We need to support initialized core for the ha_storage case
 
 	init, err := c.Initialized(ctx)
 	if err != nil {
 		return false, errwrap.Wrapf("failed to check if core is initialized: {{err}}", err)
 	}
-	if init {
+
+	_, isRaftStorage := c.underlyingPhysical.(*raft.RaftBackend)
+
+	// Prevent join from happening if we're using raft for storage and
+	// it has already been initialized.
+	if init && isRaftStorage {
 		return true, nil
 	}
 
+	if isRaftStorage && !c.Sealed() {
+		c.logger.Error("node must be sealed before joining")
+		return false, errors.New("node must be sealed before joining")
+	}
```

Comment:
Is this necessary since we just confirmed we are not initialized?

Reply:
Good catch, should be good to remove.

Reply:
I think the proper check should actually be `if isRaftHA && c.Sealed() { return false, errors.New("node must be unsealed before joining") }`

Reply:
Still think this branch is impossible to hit. Since we just confirmed we are not initialized there is no way we can be unsealed.